pkg/policy/validate.go

```diff
@@ -89,6 +95,97 @@ func Validate(p kyverno.ClusterPolicy) error {
 	return nil
 }
 
+func validateMatchExcludeConflict(rule kyverno.Rule) error {
+
+	excludeRoles := make(map[string]bool)
+	for _, role := range rule.ExcludeResources.UserInfo.Roles {
+		excludeRoles[role] = true
+	}
+
+	excludeClusterRoles := make(map[string]bool)
+	for _, clusterRoles := range rule.ExcludeResources.UserInfo.ClusterRoles {
+		excludeClusterRoles[clusterRoles] = true
+	}
+
+	excludeSubjects := make(map[string]bool)
+	for _, subject := range rule.ExcludeResources.UserInfo.Subjects {
+		subjectRaw, _ := json.Marshal(subject)
+		excludeSubjects[string(subjectRaw)] = true
+	}
+
+	excludeKinds := make(map[string]bool)
+	for _, kind := range rule.ExcludeResources.ResourceDescription.Kinds {
+		excludeKinds[kind] = true
+	}
+
+	excludeNamespaces := make(map[string]bool)
+	for _, namespace := range rule.ExcludeResources.ResourceDescription.Namespaces {
+		excludeNamespaces[namespace] = true
+	}
+
+	excludeMatchExpressions := make(map[string]bool)
+	if rule.ExcludeResources.ResourceDescription.Selector != nil {
+		for _, matchExpression := range rule.ExcludeResources.ResourceDescription.Selector.MatchExpressions {
+			matchExpressionRaw, _ := json.Marshal(matchExpression)
+			excludeMatchExpressions[string(matchExpressionRaw)] = true
+		}
+	}
+
+	for _, role := range rule.MatchResources.UserInfo.Roles {
+		if excludeRoles[role] {
+			return fmt.Errorf("excluding role '%v' while also matching it - please remove from both match and exclude", role)
+		}
+	}
+
+	for _, clusterRole := range rule.MatchResources.UserInfo.ClusterRoles {
+		if excludeClusterRoles[clusterRole] {
+			return fmt.Errorf("excluding cluster role '%v' while also matching it - please remove from both match and exclude", clusterRole)
+		}
+	}
+
+	for _, subject := range rule.MatchResources.UserInfo.Subjects {
+		subjectRaw, _ := json.Marshal(subject)
+		if excludeSubjects[string(subjectRaw)] {
+			return fmt.Errorf("excluding subject '%v' while also matching it - please remove from both match and exclude", string(subjectRaw))
+		}
+	}
+
+	if rule.MatchResources.ResourceDescription.Name != "" {
+		if rule.MatchResources.ResourceDescription.Name == rule.ExcludeResources.ResourceDescription.Name {
```

Comment:
Do we need to check the name of the resource? It's possible to have the same pod name, say "test-pod" in different namespaces.

Reply:
If there are 2 pods with same name in different name space:
Currently we can write:
```
match:
	resources:
		name: test-pod
		kinds:
			-Pod
exclude:
	resources:
		name: test-pod
		kinds:
			-Pod
		namespace:
			-namespace1
```
With this PR user will be forced to write:
```
match:
	resources:
		name: test-pod
		kinds:
			-Pod
exclude:
	resources:
		namespace:
			-namespace1
```

Reply:
If I write the following policy, it will be rejected:
````
match:
  resources:
    name: test-pod
    kinds:
      - Pod
    namespaces:
      - namespace1
exclude:
  resources:
    name: test-pod
    kinds:
      - Pod
    namespaces:
      - namespace2
````

Reply:
The above exclude block will never apply since its only matching pod from namespace1

Lets assume - this is what you meant:
```
match:
  resources:
    name: test-pod
    kinds:
      - Pod
    namespaces:
      - namespace1
      - namespace2
exclude:
  resources:
    name: test-pod
    kinds:
      - Pod
    namespaces:
      - namespace2
```
This will be rejected since its logically equivalent to this
```
match:
  resources:
    name: test-pod
    kinds:
      - Pod
    namespaces:
      - namespace1
```

Reply:
After further analysis the following policy will be rejected even though its valid

```
match:
  resources:
    namespaces:
      - namespace1
      - namespace2
exclude:
  resources:
    kinds:
      - Pod
    namespaces:
      - namespace2
```

Maybe we should reconsider this change